solidity/contracts/middleware/liquidity-layer/adapters/CctpAdapter.sol

```diff
@@ -0,0 +1,185 @@
+    event DomainAdded(uint32 indexed hyperlaneDomain, uint32 circleDomain);
+
+    /**
+     * @param _owner The new owner.
+     * @param _tokenMessenger The TokenMessenger contract.
+     * @param _token The token address.
+     * @param _tokenSymbol The token symbol.
+     * @param _gasAmount The amount of gas required to process a CCTP transfer.
+     * @param _mailbox The address of the mailbox contract.
+     * @param _interchainGasPaymaster The address of the interchain gas paymaster contract.
+     * @param _interchainSecurityModule The address of the interchain security module contract.
+     */
+    function initialize(
+        address _owner,
+        address _tokenMessenger,
+        address _token,
+        string calldata _tokenSymbol,
+        uint256 _gasAmount,
+        address _mailbox,
+        address _interchainGasPaymaster,
+        address _interchainSecurityModule
+    ) external initializer {
+        __HyperlaneConnectionClient_initialize(
+            _mailbox,
+            _interchainGasPaymaster,
+            _interchainSecurityModule,
+            _owner
+        );
+
+        tokenMessenger = ITokenMessenger(_tokenMessenger);
+        token = _token;
+        tokenSymbol = _tokenSymbol;
+        gasAmount = _gasAmount;
+    }
+
+    /**
+     * @notice Transfers `_amount` token to `_recipientAddress` on `_destinationDomain` chain.
+     * @param _destinationDomain The identifier of the destination chain.
+     * @param _recipientAddress The address of the recipient on the destination chain.
+     * @param _amount The amount of tokens to transfer.
+     * @return messageId The identifier of the dispatched message.
+     */
+    function transferRemote(
+        uint32 _destinationDomain,
+        bytes32 _recipientAddress,
+        uint256 _amount
+    ) external payable override returns (bytes32 messageId) {
+        _mustHaveRemoteRouter(_destinationDomain);
+        uint32 _circleDomain = hyperlaneDomainToCircleDomain[
+            _destinationDomain
+        ];
+
+        IERC20(token).transferFrom(msg.sender, address(this), _amount);
+        IERC20(token).approve(address(tokenMessenger), _amount);
+
+        uint64 _nonce = tokenMessenger.depositForBurn(
+            _amount,
+            _circleDomain,
+            _recipientAddress,
+            token
+        );
+
+        emit BridgedToken(_nonce);
+
+        bytes memory _message = abi.encode(
+            _recipientAddress, // The "user" recipient
+            _amount, // The amount of the tokens sent over the bridge
+            TypeCasts.addressToBytes32(msg.sender),
+            _nonce,
+            tokenSymbol
+        );
+
+        messageId = _dispatchWithGas(
+            _destinationDomain,
+            _message,
+            gasAmount,
+            msg.value,
+            msg.sender
+        );
+
+        emit SentTransferRemote(_destinationDomain, _recipientAddress, _amount);
+    }
+
+    // token transfer is already handled by the CctpIsm
```

Comment:
where is the cctpISM file and why can't be do token transfers here instead?

Reply:
CctpIsm file is in a separate repo as of now (https://github.com/Accio-labs/permissionless-cctp/blob/main/apps/contracts/contracts/CctpIsm.sol), but I'll be merging it with this CctpAdapter file as per Nam's suggestion in the comment above.